sycl/unittests/scheduler/HostTaskAndBarrier.cpp

```diff
@@ -0,0 +1,272 @@
+  sycl::event HTEvent = AddTask(TestCGType::HOST_TASK);
+  EventImplPtr HostTaskEventImpl = sycl::detail::getSyclObjImpl(HTEvent);
+  auto HostTaskWaitList = HostTaskEventImpl->getWaitList();
+  EXPECT_EQ(HostTaskWaitList.size(), 0u);
+  EXPECT_EQ(HostTaskEventImpl->isEnqueued(), true);
+
+  sycl::event BarrierEvent = AddTask(TestCGType::BARRIER);
+  EventImplPtr BarrierEventImpl = sycl::detail::getSyclObjImpl(BarrierEvent);
+  auto BarrierWaitList = BarrierEventImpl->getWaitList();
+  ASSERT_EQ(BarrierWaitList.size(), 1u);
+  EXPECT_EQ(BarrierWaitList[0], HostTaskEventImpl);
+  EXPECT_EQ(BarrierEventImpl->isEnqueued(), false);
+
+  sycl::event KernelEvent = AddTask(TestCGType::KERNEL_TASK);
+  EventImplPtr KernelEventImpl = sycl::detail::getSyclObjImpl(KernelEvent);
+  auto KernelWaitList = KernelEventImpl->getWaitList();
+  ASSERT_EQ(KernelWaitList.size(), 1u);
+  EXPECT_EQ(KernelWaitList[0], BarrierEventImpl);
+  EXPECT_EQ(KernelEventImpl->isEnqueued(), false);
+
+  MainLock.unlock();
+  QueueDevImpl->wait();
+}
+
+TEST_F(BarrierHandlingWithHostTask, HostTaskKernelBarrier) {
+  sycl::event HTEvent = AddTask(TestCGType::HOST_TASK);
+  EventImplPtr HostTaskEventImpl = sycl::detail::getSyclObjImpl(HTEvent);
+  auto HostTaskWaitList = HostTaskEventImpl->getWaitList();
+  EXPECT_EQ(HostTaskWaitList.size(), 0u);
+  EXPECT_EQ(HostTaskEventImpl->isEnqueued(), true);
+
+  sycl::event KernelEvent = AddTask(TestCGType::KERNEL_TASK);
+  EventImplPtr KernelEventImpl = sycl::detail::getSyclObjImpl(KernelEvent);
+  auto KernelWaitList = KernelEventImpl->getWaitList();
+  ASSERT_EQ(KernelWaitList.size(), 0u);
+  EXPECT_EQ(KernelEventImpl->isEnqueued(), true);
+
+  sycl::event BarrierEvent = AddTask(TestCGType::BARRIER);
+  EventImplPtr BarrierEventImpl = sycl::detail::getSyclObjImpl(BarrierEvent);
+  auto BarrierWaitList = BarrierEventImpl->getWaitList();
+  ASSERT_EQ(BarrierWaitList.size(), 1u);
+  EXPECT_EQ(BarrierWaitList[0], HostTaskEventImpl);
+  EXPECT_EQ(BarrierEventImpl->isEnqueued(), false);
+
+  MainLock.unlock();
+  QueueDevImpl->wait();
+}
+
+TEST_F(BarrierHandlingWithHostTask, BarrierHostTaskKernel) {
+  sycl::event BarrierEvent = AddTask(TestCGType::BARRIER);
+  EventImplPtr BarrierEventImpl = sycl::detail::getSyclObjImpl(BarrierEvent);
+  auto BarrierWaitList = BarrierEventImpl->getWaitList();
+  ASSERT_EQ(BarrierWaitList.size(), 0u);
+  EXPECT_EQ(BarrierEventImpl->isEnqueued(), true);
+
+  sycl::event HTEvent = AddTask(TestCGType::HOST_TASK);
+  EventImplPtr HostTaskEventImpl = sycl::detail::getSyclObjImpl(HTEvent);
+  auto HostTaskWaitList = HostTaskEventImpl->getWaitList();
+  ASSERT_EQ(HostTaskWaitList.size(), 0u);
+  EXPECT_EQ(HostTaskEventImpl->isEnqueued(), true);
+
+  sycl::event KernelEvent = AddTask(TestCGType::KERNEL_TASK);
+  EventImplPtr KernelEventImpl = sycl::detail::getSyclObjImpl(KernelEvent);
+  auto KernelWaitList = KernelEventImpl->getWaitList();
+  ASSERT_EQ(KernelWaitList.size(), 0u);
```

Comment:
Do I understand correctly that the barrier semantics is guaranteed by the backend for the kernel task, so there is not dependency on barrier?

I suggest adding in-order queue test case with code comments describing code logic.